File: balboa-http/src/main/java/com/socrata/balboa/server/rest/EntitiesRest.java
package com.socrata.balboa.server.rest;

import com.socrata.balboa.metrics.data.DataStore;
import com.socrata.balboa.metrics.data.DataStoreFactory;
import org.codehaus.jackson.map.ObjectMapper;
import org.codehaus.jackson.map.SerializationConfig;
import org.codehaus.jackson.map.annotate.JsonSerialize;

import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import java.io.IOException;
import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.regex.Pattern;

@Path("/entities")
public class EntitiesRest
{
    @GET
    @Produces("application/json")
    public Response index(
            @QueryParam("filter") String filter,
            @DefaultValue("-1") @QueryParam("limit") int limit
    ) throws IOException
    {
        throw new UnsupportedOperationException("Getting the list of entities over http is a really bad idea. Contact your System Administrator");
    }

}

File: balboa-http/src/main/scala/com/socrata/balboa/server/rest/EntitiesRest.scala
package com.socrata.balboa.server.rest

import java.util.regex.Pattern

import com.socrata.balboa.metrics.data.DataStoreFactory
import com.socrata.balboa.server.{Error, ResponseWithType}
import com.socrata.balboa.server.ResponseWithType._
import org.codehaus.jackson.map.annotate.JsonSerialize
import org.codehaus.jackson.map.{ObjectMapper, SerializationConfig}
import org.eclipse.jetty.http.HttpStatus.BAD_REQUEST_400
import org.scalatra.{BadRequest, Ok, Params}

import scala.collection.JavaConverters._
import scala.util.Try

// scalastyle:off return

object EntitiesRest {
  val dataStore = DataStoreFactory.get

  def apply(params: Params): ResponseWithType = {
    val predicate: (String => Boolean) =
      params.get("filter").map(Pattern.compile)
        .map { pattern => (str: String) => pattern.matcher(str).matches }.getOrElse(_ => true)

    val it = dataStore.entities().asScala.filter(predicate)
    // backwards-compatibility: -1 == no limit
    val limitString = params.getOrElse("limit", "-1")
    val limit = Try(limitString.toInt).recover({
      case err: NumberFormatException =>
        return ResponseWithType(json,
          BadRequest(Error(
            BAD_REQUEST_400,
            "Unable to parse limit as int")))
    }).get

    val limitedIt = if(limit != -1) it.take(limit) else it

    val mapper = new ObjectMapper
    mapper.configure(SerializationConfig.Feature.INDENT_OUTPUT, true)
    mapper.getSerializationConfig.withSerializationInclusion(JsonSerialize.Inclusion.NON_NULL)

    ResponseWithType(json, Ok(mapper.writeValueAsString(limitedIt.asJava)))
  }
}
